### PR TITLE
1544/ami overrides + unit form refactor

### PR DESCRIPTION
**Pull request by emilyjablonski**
# Pull Request Template

## Issue

Addresses #1706

- [x] This change addresses the issue in full
- [ ] This change addresses only certain aspects of the issue
- [ ] This change is a dependency for another issue
- [ ] This change has a dependency from another issue

## Description

Adds the ability to create ami chart overrides in listings management, and also due to lots of state issues refactors the UnitForm component.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Prototype/POC (not to merge)
- [x] This change is a refactor/address technical debt
- [ ] This change requires a documentation update
- [ ] This change requires a SQL Script

## How Can This Be Tested/Reviewed?

Create a new listing. Set an ami chart. Set an ami percentage from the dropdown (will pulls unique ami percentages from the selected dropdown). The table should populate with the ami chart values. Change one of the values (creating an override). Save the unit. Re-open the unit before saving the listing and ensure the override is still there. Change the percentage and the chart and ensure the override isn't sticky.

Save a listing with a unit with an overrides and view the details page for that unit and ensure the only value saved as an override is the one(s) that was overridden and that the default values aren't duplicated.

Ensure Copy and New works and brings overrides with it. Ensure Save and New and Save and Exit work as well.

Open an existing seeded listing to edit it and ensure the ami values populate.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have reviewed the changes in a desktop view
- [x] I have reviewed the changes in a mobile view
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have assigned reviewers
- [x] I have updated the changelog to include a description of my changes

### CODE REVIEW

**Review by seanmalbert — COMMENTED** [review]
@emilyjablonski ,
This is definitely looking better than where it was and it's much easier to follow the event loop . I have a couple of question/consideration comments on the code.

Other than that, I have a general question of how should we handle a user entering in a percentage that is not in the chart? Let's say that a user selects a chart and then adds 50, which has a match in the chart, but then changes it to 45, which doesn't. On the form now, the fields don't clear out, but if I edit an entry, save it and reopen, I will only see the value I added. I think we should either clear out the fields if they add a percentage that isn't in the chart, or we update the percentage field to be a select based off of the available, unique percentages in the chart.

File: sites/partners/src/listings/PaperListingForm/UnitForm.tsx [hunk 1]
```diff
@@ -53,49 +56,93 @@ const UnitForm = ({ onSubmit, onClose, units, currentTempId }: UnitFormProps) =>
...
+      })
+      const amiChartData = thisAmiChart.items
+        .filter((item) => item.percentOfAmi === parseInt(defaultAmiPercentage ?? amiPercentage))
+        .sort(function (a, b) {
+          return a.householdSize - b.householdSize
+        })
+      setCurrentAmiChart(amiChartData)
+      amiChartData.forEach((amiValue, index) => {
+        setValue(`maxIncomeHouseholdSize${index + 1}`, amiValue.income.toString())
+      })
```
[the remark this answers is not in the document]
Comment (emilyjablonski) [reply sites/partners/src/listings/PaperListingForm/UnitForm.tsx]: Makes sense, I changed amiPercentage to be a select based on the available percentages in the selected chart and am only fetching new charts on chart change

File: sites/partners/src/listings/PaperListingForm/UnitForm.tsx [hunk 2]
```diff
@@ -124,27 +171,48 @@ const UnitForm = ({ onSubmit, onClose, units, currentTempId }: UnitFormProps) =>
...
-    const current = units.find((unit) => unit.tempId === tempId)
-    if (current) {
-      onSubmit({ ...formData, id: current.id, tempId: current.tempId })
-    } else {
-      onSubmit({ ...formData, id: undefined, tempId: units.length + 1 })
-    }
-    setTempId(null)
+    const existingId = units.filter((unit) => unit.tempId === defaultUnit?.tempId)[0]?.tempId
+    const nextId = units.length + 1
+
```
[the remark this answers is not in the document]
Comment (emilyjablonski) [reply sites/partners/src/listings/PaperListingForm/UnitForm.tsx]: Passing two new props for the IDs and removed units as a prop

File: sites/partners/src/listings/PaperListingForm/UnitForm.tsx [hunk 3]
```diff
@@ -317,9 +411,31 @@ const UnitForm = ({ onSubmit, onClose, units, currentTempId }: UnitFormProps) =>
...
+          <GridCell>
+            <table>
+              <thead>
+                <tr>
+                  <th>{t("listings.householdSize")}</th>
+                  <th>{t("listings.maxAnnualIncome")}</th>
+                </tr>
+              </thead>
+              <tbody>{getAmiChartTableData()}</tbody>
+            </table>
```
[the remark this answers is not in the document]
Comment (emilyjablonski) [reply sites/partners/src/listings/PaperListingForm/UnitForm.tsx]: Added styles to make it match

File: sites/partners/src/listings/PaperListingForm/UnitForm.tsx [hunk 4]
```diff
@@ -124,27 +202,48 @@ const UnitForm = ({ onSubmit, onClose, units, currentTempId }: UnitFormProps) =>
       delete data.unitType
     }
 
-    const formData = {
+    // Only keep overrides so we're not duplicating existing ami data
+    ;[...Array(maxAmiHouseholdSize)].forEach((_, index) => {
+      const existingChartValue = currentAmiChart
+        .filter((item: AmiChartItem) => item.householdSize === index + 1)
+        .filter((item: AmiChartItem) => item.percentOfAmi === parseInt(amiPercentage))[0]
```
Comment (seanmalbert) [inline sites/partners/src/listings/PaperListingForm/UnitForm.tsx:209]: I didn't notice this before, but can you make this a single filter?

File: sites/partners/src/listings/PaperListingForm/index.tsx [hunk 5]
```diff
@@ -33,6 +33,7 @@ import {
   PaperApplication,
   PaperApplicationCreate,
   ListingReviewOrder,
+  UnitAmiChartOverrideCreate,
```
Comment (seanmalbert) [inline sites/partners/src/listings/PaperListingForm/index.tsx:36]: I don't think this is used now.

File: sites/partners/src/listings/PaperListingForm/UnitForm.tsx [hunk 6]
```diff
@@ -53,49 +59,121 @@ const UnitForm = ({ onSubmit, onClose, units, currentTempId }: UnitFormProps) =>
...
-    const unit = units.filter((unit) => unit.tempId === tempId)[0]
-    setCurrent(unit)
-    reset({
-      ...unit,
-      rentType: getRentType(unit),
-      status: UnitStatus.available,
+    void resetDefaultValues()
+  }, [])
+
+  const fetchAmiChart = async (defaultChartID?: string, defaultAmiPercentage?: string) => {
```
Comment (seanmalbert) [inline sites/partners/src/listings/PaperListingForm/UnitForm.tsx:126]: defaultAmiPercentage isn't used

File: sites/partners/src/listings/PaperListingForm/UnitForm.tsx [hunk 7]
```diff
@@ -173,6 +272,25 @@ const UnitForm = ({ onSubmit, onClose, units, currentTempId }: UnitFormProps) =>
     })
   }, [amiCharts, unitPriorities, unitTypes])
 
+  useEffect(() => {
+    if (defaultUnit) {
+      setValue("amiChart.id", defaultUnit.amiChart?.id)
+      setValue("priorityType.id", defaultUnit.priorityType?.id)
+      setValue("unitType.id", defaultUnit.unitType?.id)
+    }
+  }, [options])
```
Comment (seanmalbert) [inline sites/partners/src/listings/PaperListingForm/UnitForm.tsx:281]: The linter is complaining about the dependency arrays for this and two others. Does updating them cause issues?

**Review by seanmalbert — APPROVED** [review]
@emilyjablonski ,
This looks good and seems to be working well. There's some minor cleanup and then I think you can merge.

**Pull request merged**